### PR TITLE
Bump go, rules_go, and bazel_gazelle

### PR DESCRIPTION
**Pull request by sudowork**
Certain vulnerability scanners will flag bazelisk for being built with an outdated golang toolchain. To remediate this, update to the latest version of the go toolchain (1.19.3). Upgrade to the latest major version since 1.16 is EOL.

Also upgrades `rules_go` to `0.36.0` and `bazel_gazelle` to `0.28.0`. At the same time, removes the workaround to get output binaries placed into the appropriate directory since it's no longer necessary.

### CODE REVIEW

**Comment by sudowork** [comment]
Output of running `./build.sh` and `go test -v` locally:

```
$ ./build.sh
INFO: Analyzed 6 targets (1 packages loaded, 8846 targets configured).
INFO: Found 6 targets...
INFO: Elapsed time: 131.644s, Critical Path: 106.73s
INFO: 62 processes: 2 internal, 60 darwin-sandbox.
INFO: Build completed successfully, 62 total actions

== Bazelisk binaries are ready ==
-r-xr-xr-x  1 kgao  staff    10M Dec  5 07:55 bin/bazelisk-darwin
-r-xr-xr-x  1 kgao  staff   5.3M Dec  5 07:55 bin/bazelisk-darwin-amd64
-r-xr-xr-x  1 kgao  staff   5.1M Dec  5 07:55 bin/bazelisk-darwin-arm64
-r-xr-xr-x  1 kgao  staff   4.9M Dec  5 07:55 bin/bazelisk-linux-amd64
-r-xr-xr-x  1 kgao  staff   4.8M Dec  5 07:55 bin/bazelisk-linux-arm64
-r-xr-xr-x  1 kgao  staff   7.2M Dec  5 07:55 bin/bazelisk-windows-amd64.exe
bin/bazelisk-darwin:            Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64]
bin/bazelisk-darwin (for architecture x86_64):	Mach-O 64-bit executable x86_64
bin/bazelisk-darwin (for architecture arm64):	Mach-O 64-bit executable arm64
bin/bazelisk-darwin-amd64:      Mach-O 64-bit executable x86_64
bin/bazelisk-darwin-arm64:      Mach-O 64-bit executable arm64
bin/bazelisk-linux-amd64:       ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=redacted, stripped
bin/bazelisk-linux-arm64:       ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, Go BuildID=redacted, stripped
bin/bazelisk-windows-amd64.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows

== Bazelisk version output ==
Before releasing, make sure that this is the correct version string:
Bazelisk version: 4c5100d75c991757b51b6e5b9b65925b35bfe450

== NPM releases ==
After testing, publish to NPM via these commands:
$ npm login --registry https://wombat-dressing-room.appspot.com
$ ./bazelisk run --config=release //:npm_package.publish -- --access=public --tag latest --registry https://wombat-dressing-room.appspot.com
```

```
$ go test -v
=== RUN   TestResolveVersion
--- PASS: TestResolveVersion (0.00s)
=== RUN   TestResolvePatchVersion
--- PASS: TestResolvePatchVersion (0.00s)
=== RUN   TestResolveLatestRcVersion
--- PASS: TestResolveLatestRcVersion (0.00s)
=== RUN   TestResolveLatestRcVersion_WithFullRelease
--- PASS: TestResolveLatestRcVersion_WithFullRelease (0.00s)
=== RUN   TestResolveLatestVersion_TwoLatestVersionsDoNotHaveAReleaseYet
--- PASS: TestResolveLatestVersion_TwoLatestVersionsDoNotHaveAReleaseYet (0.00s)
=== RUN   TestResolveLatestVersion_ShouldOnlyReturnStableReleases
--- PASS: TestResolveLatestVersion_ShouldOnlyReturnStableReleases (0.00s)
=== RUN   TestResolveLatestVersion_ShouldFailIfNotEnoughReleases
--- PASS: TestResolveLatestVersion_ShouldFailIfNotEnoughReleases (0.00s)
=== RUN   TestResolveLatestVersion_GCSIsDown
--- PASS: TestResolveLatestVersion_GCSIsDown (3.47s)
=== RUN   TestResolveLatestVersion_GitHubIsDown
--- PASS: TestResolveLatestVersion_GitHubIsDown (0.00s)
=== RUN   TestAcceptRollingReleaseName
--- PASS: TestAcceptRollingReleaseName (0.00s)
=== RUN   TestResolveLatestRollingRelease
--- PASS: TestResolveLatestRollingRelease (0.00s)
=== RUN   TestAcceptFloatingReleaseVersions
--- PASS: TestAcceptFloatingReleaseVersions (0.00s)
PASS
ok  	github.com/bazelbuild/bazelisk	3.725s
```

**Comment by sudowork** [comment]
Dug into why the CI builds were failing and believe my additional two commits have resolved the issue. Should be ready for review now. Let me know if there's anything you'd like me to address.

**Comment by sudowork** [comment]
Sorry for the ping, as I'm not sure what the proper etiquette/process for flagging something for review is here. Is this something that can be looked at @fweikert @philwo?

**Comment by fmeum** [comment]
@sudowork Feel free to apply this patch on top to update to Go 1.19.4 and also keep the rules_go Bazel repo in sync with the Go dependency on it:

```patch
diff --git a/WORKSPACE b/WORKSPACE
index bf24ac1..8036451 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -24,15 +24,15 @@ load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies", "go_repository")
 go_repository(
     name = "com_github_bazelbuild_rules_go",
     importpath = "github.com/bazelbuild/rules_go",
-    sum = "h1:KViqR7qKXwz+LrNdIauCDU21kneCk+4DnYjpvlJwH50=",
-    version = "v0.27.0",
+    sum = "h1:hKvFRTyZEoHeIjPe/Kd8RL9cP9s3hoeMlO8vywdXSfY=",
+    version = "v0.36.0",
 )
 
 go_repository(
     name = "com_github_hashicorp_go_version",
     importpath = "github.com/hashicorp/go-version",
-    sum = "h1:McDWVJIU/y+u1BRV06dPaLfLCaT7fUTJLp5r04x7iNw=",
-    version = "v1.3.0",
+    sum = "h1:feTTfFNnjP967rlCxM/I9g701jU+RN74YKx2mOkIeek=",
+    version = "v1.6.0",
 )
 
 go_repository(
@@ -44,15 +44,14 @@ go_repository(
 
 go_repository(
     name = "com_github_bgentry_go_netrc",
-    importpath = "github.com/bgentry/go-netrc/netrc",
-    urls = ["https://github.com/bgentry/go-netrc/archive/9fd32a8.zip"],
-    type = "zip",
-    strip_prefix = "go-netrc-9fd32a8b3d3d3f9d43c341bfe098430e07609480/netrc",
+    importpath = "github.com/bgentry/go-netrc",
+    sum = "h1:xDfNPAt8lFiC1UJrqV3uuy861HCTo708pDMbjHHdCas=",
+    version = "v0.0.0-20140422174119-9fd32a8b3d3d",
 )
 
 go_rules_dependencies()
 
-go_register_toolchains(version = "1.19.3")
+go_register_toolchains(version = "1.19.4")
 
 gazelle_dependencies()
 
diff --git a/go.mod b/go.mod
index b433331..34fd468 100644
--- a/go.mod
+++ b/go.mod
@@ -1,9 +1,9 @@
 module github.com/bazelbuild/bazelisk
 
-go 1.15
+go 1.19
 
 require (
-	github.com/bazelbuild/rules_go v0.34.0
+	github.com/bazelbuild/rules_go v0.36.0
 	github.com/bgentry/go-netrc v0.0.0-20140422174119-9fd32a8b3d3d
 	github.com/hashicorp/go-version v1.6.0
 	github.com/mitchellh/go-homedir v1.1.0
diff --git a/go.sum b/go.sum
index e5ef5fd..f71cf96 100644
--- a/go.sum
+++ b/go.sum
@@ -1,5 +1,5 @@
-github.com/bazelbuild/rules_go v0.34.0 h1:cmObMtgIOaEU944SqXtJ9DnlS8IPGGa7pdRnsrpQzXM=
-github.com/bazelbuild/rules_go v0.34.0/go.mod h1:MC23Dc/wkXEyk3Wpq6lCqz0ZAYOZDw2DR5y3N1q2i7M=
+github.com/bazelbuild/rules_go v0.36.0 h1:hKvFRTyZEoHeIjPe/Kd8RL9cP9s3hoeMlO8vywdXSfY=
+github.com/bazelbuild/rules_go v0.36.0/go.mod h1:TMHmtfpvyfsxaqfL9WnahCsXMWDMICTw7XeK9yVb+YU=
 github.com/bgentry/go-netrc v0.0.0-20140422174119-9fd32a8b3d3d h1:xDfNPAt8lFiC1UJrqV3uuy861HCTo708pDMbjHHdCas=
 github.com/bgentry/go-netrc v0.0.0-20140422174119-9fd32a8b3d3d/go.mod h1:6QX/PXZ00z/TKoufEY6K/a0k6AhaJrQKdFe6OfVXsa4=
 github.com/hashicorp/go-version v1.6.0 h1:feTTfFNnjP967rlCxM/I9g701jU+RN74YKx2mOkIeek=

```

**Comment by sudowork** [comment]
Applied the patch, but left the `com_github_bgentry_go_netrc` `go_repository` as a zip download since gazelle was not producing a BUILD file for the package as defined in the patch (perhaps because the `.go` files are not in the root?):
```
ERROR: /Users/kgao/dev/bazelisk/httputil/BUILD:7:11: no such package '@com_github_bgentry_go_netrc//': BUILD file not found in directory '' of external repository @com_github_bgentry_go_netrc. Add a BUILD file to a directory to mark it as a package. and referenced by '//httputil:go_default_library'
ERROR: Analysis of target '//:go_default_library' failed; build aborted: 
INFO: Elapsed time: 1.053s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (26 packages loaded, 1370 targets configured)
    Fetching @local_config_xcode; Building xcode-locator
    Fetching @com_github_mitchellh_go_homedir; fetching
```

**Comment by sudowork** [comment]
@fmeum , I've bumped up the version of rules_go to 0.37.0 since that released on 2022-12-06. Is there anything else you'd like me to do before this can be reviewed/merged?

**Comment by fmeum** [comment]
@sudowork Looks good, thanks for merging in my changes! I don't have write access to this repo, so I can't review or merge this PR, but if I could, I would :)

**Comment by sudowork** [comment]
Aha oops that was my mistake. Thanks for your patch and review @fmeum! I think I meant to tag @fweikert.

**Review by fweikert — APPROVED** [review]

**Comment by fweikert** [comment]
Nice, thank you so much! I'll check whether we can now build releases on M1 Macs (which previously failed due to the old rules_go version).

**Pull request merged**
